### PR TITLE
docs: use custom diff code blocks

### PR DESCRIPTION
**Pull request by Josh-Cena**
## PR Checklist

-   [ ] Addresses an existing open issue: fixes #000
-   [ ] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [ ] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

The diff makes the code not directly copyable, which is not something you want for tutorial examples.

See related discussions at:
- https://github.com/facebook/react-native-website/pull/2984
- https://github.com/facebook/docusaurus/pull/6833
- https://github.com/facebook/docusaurus/pull/4973

### CODE REVIEW

**Review by bradzacher — COMMENTED** [review]
Is this something we can build a docusaurus plugin for? To add the +/- in via css so the  content is copyable?

I personally like diffs as it clearly shows what changes should be made which helps tell the story of how to configure things, IMO

**Comment by Josh-Cena** [comment]
We can, absolutely. but it would still mean showing two competing lines and when users copy, they have to delete the "-" line.

**Comment by Josh-Cena** [comment]
I don't know what's going on with the Netlify preview, though—it's not showing the latest content😕

**Comment by armano2** [comment]
> I personally like diffs as it clearly shows what changes should be made which helps tell the story of how to configure things

i do agree with that, if we want to make file copyable, maybe `user-select: none;` could be used instead? for `+` and `-`

```css
.token.prefix {
    user-select: none;
}
```

**Comment by armano2** [comment]
> I personally like diffs as it clearly shows what changes should be made which helps tell the story of how to configure things

i do agree with that, if we want to make file copyable, maybe `user-select: none;` could be used instead? for `+` and `-`

```css
.token.prefix {
    user-select: none;
}
```

**Comment by bradzacher** [comment]
https://jsfiddle.net/fp2jrnmv/
Just to show a quick-and-dirty POC of this working in a fiddle.
There is one small issue - we can't prevent copying empty newlines[1] - but it's otherwise perfect.


[1] copying the above example yields:
```
  module.exports = {
    root: true,
    parser: '@typescript-eslint/parser',
    plugins: [
      '@typescript-eslint',
    ],
    extends: [


     'airbnb-typescript',
    ],
  };
```

Another option (or an additional option) could be to use JS to add a copy button which completely deletes the deleted lines from the copied contents.

**Comment by armano2** [comment]
> There is one small issue - we can't prevent copying empty newlines[1] 

```css
.language-diff .token-line > .deleted + br
```

**Comment by bradzacher** [comment]
@armano2 - ah I missed the `br`s!

<details>
<summary>The result of the copy paste with armano's updated css</summary>

https://jsfiddle.net/fmt0h31z/

```
  module.exports = {
    root: true,
    parser: '@typescript-eslint/parser',
    plugins: [
      '@typescript-eslint',
    ],
    extends: [
     'airbnb-typescript',
    ],
  }
```

</details>

This is **exactly** what we want - you can copy paste the true snipped AND we get to show the changes to tell a story.
I love it!

**Comment by Josh-Cena** [comment]
Great, I'll implement this and you can see the effect (hopefully—if the deploy preview updates)

**Comment by bradzacher** [comment]
> Also, what do we do about the copy button? Should we hide it entirely, or do you want it to have custom behavior as well?

If it's not too difficult - having it do the same thing (copy the "intended" code) would be great!
Otherwise I'm happy for it to be hidden if it's too hard to customise.

File: packages/website/docusaurusConfig.ts [hunk 1]
```diff
@@ -139,6 +139,23 @@ const themeConfig: ThemeCommonConfig & AlgoliaThemeConfig = {
...
+        className: 'code-block-removed-line',
+        line: 'Remove this line',
+        block: { start: 'Removed lines start', end: 'Removed lines end' },
+      },
+      {
+        className: 'code-block-added-line',
+        line: 'Add this line',
+        block: { start: 'Added lines start', end: 'Added lines end' },
+      },
+    ],
```
Comment (Josh-Cena) [inline packages/website/docusaurusConfig.ts:158]: New Docusaurus feature to add custom metadata to lines—suits this use-case quite well.

I prefer using human-readable identifiers like "Remove this line", but if you want something that stands out more like for machines (e.g. "remove-next-line") just let me know

**Comment by Josh-Cena** [comment]
Ahhh... Customizing the copy button, even hiding it, is a bit hard at this stage. I'll make some changes in Docusaurus as well.

**Comment by Josh-Cena** [comment]
Current UI/UX is quite good IMO. We have syntax highlighting which is missed in diff code blocks. It's also more obvious that the `+`/`-` are non-text tokens because they are shifted left.

**Review by bradzacher — APPROVED** [review]
Whilst I would love if we could just use `diff` cos it's concise - i think the syntax highlighting we get from using `js` is better and paired with the new UX is well worth it.

keanubeautiful

**Comment by Josh-Cena** [comment]
Cross-posting from https://github.com/facebook/react-native-website/pull/2984—we'll probably consider making this the default behavior in Docusaurus.

**Pull request merged**